### DISCUSSION TITLE
Ability to pass a custom tsconfig for `next dev`

### DISCUSSION THREAD

**Discussion by jerrygreen** [vercel/next.js#23972] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

I'm bored of seeing `someVar is declared but its value is never read`:

<p align="center"><img src="https://user-images.githubusercontent.com/13215662/64862750-bee44400-d64c-11e9-9a90-859ada1d3b00.png" width="80%"></p>

I'm actively write and delete code, exploring some libraries' features - I don't want my dev server to stuck me into an error. Of course it's bad to leave them. Of course my IDE should highlight that - and for that reason, I want to have `noUnusedLocals` or `noUnusedParameters`, `strict` in overall. Of course there should be some pre-commit hook that will prevent me from committing this clutter into the repository. But I don’t want my dev server to throw `noUnusedLocals` and `noUnusedParameters` - I don't want it to prevent me playing with the code.

UPD. That’s okay for the default behavior, but there should be an option that these keys should be enabled by VSCode, and by linter, but not used by dev server so it won’t throw error that prevents you to do anything.

## Describe the solution you'd like

I'd like an ability of passing a tsconfig path for `next dev` command.

Something like:

```bash
next dev --tsconfig=tsconfig.devserver.json
```

While I can use the original tsconfig.json for my IDE.

## Describe alternatives you've considered

Well, I could make an issue for IDE (`vscode` doesn't support anything other than `tsconfig.json` too), but anyway - both sides should support that.

UPD. There's one: https://github.com/Microsoft/vscode/issues/12463. Still, both sides should support that.

## Additional context

Idk maybe it's possible. But I don't see it from `next dev --help`:

![image](https://user-images.githubusercontent.com/13215662/61650497-4b1d5f00-accd-11e9-89c6-ab5dca8a6d21.png)

**Comment by Timer** [comment]
Aren't these specific flags better suited for a linter like ESLint?

**Comment by jerrygreen** [comment]
@Timer yep, for both: linters and IDEs. Not for dev servers - it interrupts the development process. So need to differentiate them by choosing different configs.

P.S. There's also an issue on vscode repo - looks like it's hard to support choosing a custom config from their side: https://github.com/Microsoft/vscode/issues/12463

**Comment by Timer** [comment]
I agree right now overly strict TypeScript configurations can cause a bit of headache during development. Typically, this is solved by a separate `tsconfig.prod.json` you run over your code base using `tsc` w/ noEmit.

As for now, we're going to start working to make TypeScript errors as unobtrusive as possible before we hand out the knobs to completely disable this integration.

Please bear with us! 🙏

**Comment by gothy** [comment]
Even trying to disable this compiler error by setting _"noUnusedLocals": false_ doesn't help, it looks like some tsconfig.json setting are ignored 🤔

**Comment by fabb** [comment]
i'd like to exclude `test.tsx?` files from next compilation, but need to include the files in `tsconfig.json` so i get proper autocomplete in VSCode. any ideas how to solve this?

**Comment by eric-burel** [comment]
When using an `src` folder, an alternative could be to exclude `*.tests.ts`, `*.stories.ts` etc. in the main config, but create a new `tsconfig.json` that reinclude them, in `src/tsconfig.json`. 
This config would only be targeted at VS code.

However I am not sure if it's right. More broadly I lack the knowledge on how to check precisely which files are included in which build to debug that. The interactions between VS Code, Next, and TypeScript are unclear to the end user.

**Comment by mulholo** [comment]
> When using an src folder, an alternative could be to exclude *.tests.ts, *.stories.ts etc. in the main config, but create a new tsconfig.json that reinclude them, in src/tsconfig.json.

I followed this approach and it worked really well for me. Thanks!

**Reply by eric-burel** [reply]
It seems that it still build my stories during Next build though. If you use this approach, could you please create a story with an obvious mistake (like importing smth that doesn't exist) and tell me if `next build` still works for you? That would help me a lot

**Reply by eric-burel** [reply]
I've posted a question on SO to get more feedback on how to debug this, because it's also difficult to tell what will be included or not: https://stackoverflow.com/questions/67637255/list-files-built-by-typescript-depending-on-current-tsconfig

**Comment by danawoodman** [comment]
Is there any plans to support this in the near future?

This is causing me a lot of headaches because Next is trying to compile my test setup files in my root which means I have to have my test dependencies in `dependencies` because Heroku barfs 🤮  on `next build` when deploying 😢

**Comment by eric-burel** [comment]
Funky bug I had today:
Since Jest tests are part of the tsconfig, `test` is a valid global variable. I mistakenly misnamed a variable `isTest` and used wrongly it as `test` => the build was fine locally, but failed on the server where I don't install Jest at all.

**Comment by karlhorky** [comment]
Now that [TypeScript 4.1 is out](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#react-17-jsx-factories) with `compilerOptions.jsx` values of both `react-jsx` and `react-jsxdev`, maybe this idea of two separate configs has an additional reason to exist?

I suppose as suggested in https://github.com/facebook/create-react-app/issues/10025, it would be possible to default to `react-jsxdev` in the config file, and then in the `next build` script, mutate that to `react-jsx`...?

My question, which I also posed in the CRA issue above, is whether there are large, useful differences in functionality between `react-jsxdev` and `react-jsx`.

**Comment by eric-burel** [comment]
Hi guys, up on this, this is still problematic for us. Stories are included in `yarn build` which prevents us from making Storybook an optional dependency, consume build time for nothing, and generate unexpected bugs.

**Reply by eric-burel** [reply]
@timneutkens you turned the issue into a discussion recently, do you have a bit more feedback? The alternative to a custom config is not completely obvious

**Comment by pcottle** [comment]
Just chiming in that passing in a custom `tsconfig` for my next environment would eliminate a ton of headaches for us as well.

The issue we are dealing with is that we are sharing code between both a Firebase Cloud functions project and a Next.JS project. The code sharing is absolutely amazing -- everything is typescript, you can utilize functions and code between verticals of the app, and it all works out of the box.

The issue comes to deploying. Firebase *needs* the JS files to be emitted and can't handle `.jsx` input, so the `jsx` option is ideally `react-jsx`.

Meanwhile NextJS will revert it over to `preserve` every time we start up the NextJS site, which then breaks our deployment scripts (and tests) in firebase. I tried using babel to fix this with another build step but there are reasons why thats undesirable too.

It seems reasonable to just be able to pass in separate tsconfig to the NextJS environment 😇
